dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -25,8 +36,131 @@ namespace ErrorCodes
 {
 extern const int BAD_ARGUMENTS;
 extern const int NOT_IMPLEMENTED;
+extern const int CANNOT_INSERT_NULL_IN_ORDINARY_COLUMN;
 } // namespace ErrorCodes
 
+namespace
+{
+template <typename T>
+consteval bool checkIfSimpleNumericType()
+{
+    return std::is_integral_v<T> || std::is_floating_point_v<T>;
+}
+
+template <typename T>
+struct DecimalWithScale
+{
+    DecimalWithScale(T decimal_, Int64 scale_)
+        : decimal(decimal_)
+        , scale(scale_)
+    {
+        if constexpr (!IsDecimal<T>)
+            throw Exception("DecimalWithScale only accepts Decimal type.");
+    }
+
+    T decimal; // This T should be decimal type
+    Int64 scale;
+};
+
+using DecimalWithScale32 = DecimalWithScale<Decimal32>;
+using DecimalWithScale64 = DecimalWithScale<Decimal64>;
+using DecimalWithScale128 = DecimalWithScale<Decimal128>;
+using DecimalWithScale256 = DecimalWithScale<Decimal256>;
+
+template <typename T>
+consteval bool checkIfDecimalWithScaleType()
+{
+    return std::is_same_v<T, DecimalWithScale32> || std::is_same_v<T, DecimalWithScale64> || std::is_same_v<T, DecimalWithScale128> || std::is_same_v<T, DecimalWithScale256>;
+}
+
+// When T is Decimal, we should convert it to DecimalWithScale type
+// as we need scale value when executing the comparison operation.
+template <typename T>
+struct ActualCmpDataType
+{
+    using Type = std::conditional_t<checkIfSimpleNumericType<T>(), T, DecimalWithScale<T>>;
+};
+
+template <typename T>
+typename ActualCmpDataType<T>::Type getValue(const ColumnPtr & col_ptr, size_t idx)
+{
+    if constexpr (checkIfSimpleNumericType<T>())
+    {
+        const auto * col = static_cast<const ColumnVector<T> *>(&(*col_ptr));
```

Comment:
Can Field be used like https://github.com/pingcap/tiflash/blob/892c2cac41d95d3399bd46278a4eba44f1caa25e/dbms/src/WindowFunctions/IWindowFunction.cpp#L336-L361 ? Although it will introduce the overhead of constructing Field

Reply:
> Can Field be used like
> 
> https://github.com/pingcap/tiflash/blob/892c2cac41d95d3399bd46278a4eba44f1caa25e/dbms/src/WindowFunctions/IWindowFunction.cpp#L336-L361
> 
> ? Although it will introduce the overhead of constructing Field

I think it's better to avoid `Field` as possible as we can.